pkgs/development/python-modules/scancode-toolkit/default.nix

```diff
@@ -0,0 +1,125 @@
+{ lib
+, fetchPypi
+, buildPythonPackage
+, isPy3k
+, markupsafe
+, click
+, typecode
+, gemfileparser
+, pefile
+, fingerprints
+, spdx-tools
+, fasteners
+, pycryptodome
+, urlpy
+, dparse
+, jaraco_functools
+, pkginfo
+, debian-inspector
+, extractcode
+, ftfy
+, pyahocorasick
+, colorama
+, jsonstreams
+, packageurl-python
+, pymaven-patch
+, nltk
+, pygments
+, bitarray
+, jinja2
+, javaproperties
+, boolean-py
+, license-expression
+, extractcode-7z
+, extractcode-libarchive
+, typecode-libmagic
+, pytestCheckHook
+}:
+let
+  bitarray' = bitarray.overridePythonAttrs ({ ... }: rec {
+    pname = "bitarray";
+    version = "0.9.3";
+    src = fetchPypi {
+      inherit pname version;
+      sha256 = "b2783d74247abb2669f581a001ecaa747e4b2ff9440762dbb499ed69b83e321e";
+    };
+  });
+in
+buildPythonPackage rec {
+  pname = "scancode-toolkit";
+  version = "21.3.31";
+  disabled = !isPy3k;
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "7e0301031a302dedbb4304a91249534f3d036f84a119170b8a9fe70bd57cff95";
+  };
+
+  dontConfigure = true;
+
+  # * dparse2 is a "Temp fork for Python 2 support", but pdfminer requires
+  # Python 3, so it's "fine" to leave dparse2 unpackaged and use the "normal"
+  # version
+  # * ftfy was pinned for similar reasons (to support Python 2), but rather than
+  # packaging an older version, I figured it would be better to remove the
+  # erroneous (at least for our usage) version bound
+  postPatch = ''
+    substituteInPlace setup.cfg \
+      --replace "dparse2" "dparse" \
+      --replace "ftfy <  5.0.0" "ftfy"
+  '';
+
+  propagatedBuildInputs = [
+    markupsafe
+    click
+    typecode
+    gemfileparser
+    pefile
+    fingerprints
+    spdx-tools
+    fasteners
+    pycryptodome
+    urlpy
+    dparse
+    jaraco_functools
+    pkginfo
+    debian-inspector
+    extractcode
+    ftfy
+    pyahocorasick
+    colorama
+    jsonstreams
+    packageurl-python
+    pymaven-patch
+    nltk
+    pygments
+    bitarray'
+    jinja2
+    javaproperties
+    boolean-py
+    license-expression
+    extractcode-7z
+    extractcode-libarchive
+    typecode-libmagic
+  ];
+
+  checkInputs = [
+    pytestCheckHook
+  ];
+
+  checkPhase = ''
```

Comment:
I was unable to get tests to work if I set `preCheck = "export HOME=$(mktemp -d)";` (the tests would complain about an unwritable home), so this *is* necessary (unless there's something I overlooked).

Reply:
That happens in the `pythonImportsCheckPhase`, so we can set `$HOME` in another phase like `postInstall` and add a comment as to why.
To solve this problem completely, we should introduce `prePythonImportsCheck` or something. Alternatively, `pythonImportsCheck` should be executed after `pytestCheckPhase`.
cc @jonringer @FRidh

Reply:
should be able to do as part of `preCheck`. As I believe pythonImportsCheck is part of the `preDist` phase

Reply:
The real problem is actually addressed later: https://github.com/NixOS/nixpkgs/pull/120009#discussion_r618741635

*Importing* needs a writable home, and `preCheck` happens in between `pythonImportsCheckPhase` and `pytestCheckPhase`.

Reply:
How about we just run `preCheck` before `pythonImportsCheck` as well? https://github.com/NixOS/nixpkgs/pull/120300